### PR TITLE
Adding support for active particles

### PR DESCRIPTION
**Pull request by fearmayo**
adding support for active particles

## PR Checklist

- [ ] Code passes flake8 checker
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Hi,

we already have support for enzo-3.0 active particles. What issues did you run into this these changes solve? Do the existing test datasets we have with active particles still work correctly with these changes?

**Comment by ngoldbaum** [comment]
It looks like the answer to my second question is no, since the tests for loading active particle data are failing.

**Comment by fearmayo** [comment]
Hi Nathan,

The problem I found was that while there is support for enzo3.0 active particles there is no support for enzo-dev active particles. 

OK. Looks like I need to follow up and see what now breaks with the ap tests.

**Comment by ngoldbaum** [comment]
enzo-dev actuve particles? like for my WIP pull request to enzo? Have you continues that work? Because it’s a work-in-progress pull request we did not make the necessary changes to yt to add support for it yet.

If you have taken that work farther than I took it, great, however I think we need to wait on adding support for it here until the Enzo pull request is merged at least. Also we need to continue supporting enzo-3.0 active particles since people are doing production work with that, so you’ll need to make sure all the existing tests pass. In addition we’ll need a new test dataset and new tests for the new functionality
you want to add.

Perhaps it would be best to start a discussion about the way forward for active particles on the enzo-dev mailing list. I’m very keen to get that Enzo pull request merged and wasn’t aware that people were using the WIP code!

**Comment by fearmayo** [comment]
Yeah I pulled that WIP into my branch and it's working (there are probably a few updates required in your PR and docs of course but not too much). I have a couple of more things I need to check but in principle it works fine. I'm taking the Renaissance datasets and zooming into them with active particles and since Enzo-3.0 is being effectively discontinued I reverted to enzo-dev.

**Comment by ngoldbaum** [comment]
OK, well hopefully the path forward I outlined in my previous pull request will allow you to make progress, just to summarize here's what needs to happen on this PR before it can be merged:

- [ ] All existing tests pass
- [ ] Add new tests, including adding a new example dataset to the data archive at yt-project.org/data and new tests that use that new dataset.
- [ ] Add a small update to the docs on loading enzo data
- [ ] Enzo pull request merged, or at least have a discussion on enzo-dev about the active particle storage format so ensure it gets stabilized in a fashion that people agree on.

**Comment by matthewturk** [comment]
@fearmayo Is this PR still on your radar?

**Comment by matthewturk** [comment]
Would anyone be able to take this PR under their wing?

**Comment by fearmayo** [comment]
> Would anyone be able to take this PR under their wing?

Yes. I can restart this. The APs are in Enzo now. Give me a few weeks though as teaching prep is weighing on me right now.

**Comment by brittonsmith** [comment]
@yt-fido test this please

**Comment by brittonsmith** [comment]
Does anyone understand the Build and Test failures?

File: yt/frontends/enzo/io.py
Comment (brittonsmith) [inline yt/frontends/enzo/io.py:95]: The use of `.value` has been removed in newer versions of h5py. This could be what's causing the problem. If you switch it back to how it was, this should get things working.
```suggestion
                        np.asarray(pds.get(pn % ax)[()], dtype="=f8") for ax in "xyz"
```

File: yt/frontends/enzo/io.py
Comment (chrishavlin) [inline yt/frontends/enzo/io.py:82]: I think should this be `g.NumberOfParticles[ptype]`?

Comment (chrishavlin) [reply yt/frontends/enzo/io.py:82]: oh... hmm. Looks like I was wrong about this? doesn't look like any tests failed due to this.

Comment (chrishavlin) [reply yt/frontends/enzo/io.py:82]: oh, ya, I skimmed through too quickly. `NumberOfParticles` is not the same as `NumberOfActiveParticles`... Ignore above comments :)

File: yt/frontends/enzo/data_structures.py
Comment (brittonsmith) [inline yt/frontends/enzo/data_structures.py:377]: The failing test is because our sample data has a different name for the active particles in the file. The code below made it work for me.
```suggestion
                grid_group = handle[f"/Grid{g.id:08d}"]
                found = False
                for pname in ["Active Particles", "Particles"]:
                    if pname in grid_group:
                        found = True
                        break
                if not found:
                    raise RuntimeError(
                        "Could not find active particle group in data.")
                node = grid_group[pname]
```

File: yt/frontends/enzo/io.py
Comment (brittonsmith) [inline yt/frontends/enzo/io.py:94]:
```suggestion
                            if pds is not None:
```

File: yt/frontends/enzo/io.py
Comment (brittonsmith) [inline yt/frontends/enzo/io.py:96]:
```suggestion
                        if pds is None:
```

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/enzo/data_structures.py:384]: I know the "for-else" construct is somewhat "obscure", but this is exactly what it's supposed to be used for :)
```suggestion
                for pname in ["Active Particles", "Particles"]:
                    if pname in grid_group:
                        break
                else:
                    raise RuntimeError("Could not find active particle group in data.")
```

Comment (neutrinoceros) [inline yt/frontends/enzo/data_structures.py:384]: I know the "for-else" construct is somewhat "obscure", but this is exactly what it's supposed to be used for :)
```suggestion
                for pname in ["Active Particles", "Particles"]:
                    if pname in grid_group:
                        break
                else:
                    raise RuntimeError("Could not find active particle group in data.")
```

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/enzo/data_structures.py:393]: Trying to use better numpy idioms here
```suggestion
                        if node[ptype][field].ndim > 1:
                            self.io._array_fields[field] = (node[ptype][field][1],)
```

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/enzo/data_structures.py:376]: I can't make this a suggestion, but `h5py.File` now implements the context manager protocol. In other words:
```python
handle = h5py.File(...)
...
handle.close()
```
can (and should) be refactored into
```python
with h5py.File(...) as handle:
    ...
```

File: yt/frontends/enzo/io.py
Comment (neutrinoceros) [inline yt/frontends/enzo/io.py:91]:
```suggestion
                    elif not g.NumberOfActiveParticles[ptype]:
                        continue
                    else:
```

File: yt/frontends/enzo/io.py
Comment (neutrinoceros) [inline yt/frontends/enzo/io.py:96]:
```suggestion
                        else:
```

Comment (fearmayo) [reply yt/frontends/enzo/io.py:96]: It's not obvious to me that `else` is the right approach here. The 'if' statement is checking if 'pds' is 'None'

File: yt/frontends/enzo/data_structures.py
Comment (fearmayo) [reply yt/frontends/enzo/data_structures.py:376]: Interesting. So everything between the 'with' and the 'close is indented and in that scope? Is that how that works?

**Comment by neutrinoceros** [comment]
I've reviewed this purely with code-quality in mind. I don't know enough about what's actually going on to do an _actual_ review here. FWIW I think that actual test coverage for this is required if we're going to merge it before the yt 4.0 release

**Comment by fearmayo** [comment]
> I've reviewed this purely with code-quality in mind. I don't know enough about what's actually going on to do an _actual_ review here. FWIW I think that actual test coverage for this is required if we're going to merge it before the yt 4.0 release

Thanks very merge. Test coverage also on the way.

File: yt/frontends/enzo/io.py
Comment (neutrinoceros) [reply yt/frontends/enzo/io.py:96]: but that should only ever happen in case the `break` statement above isn't hit, correct ?
The `else` in `for` loops can be read as `if no break`. I've thought about this for a a minute and I really can't see any case where it's not equivalent, because the for loop is _always_ going to run (given the iterable is hard-coded as non-empty), so I can't see how `pds` could end up as `None` in any other scenario than "the break wasn't hit"

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/enzo/data_structures.py:376]: that's it. Basically it works as a the builtin `open` function. At the cost of adding one indentation level to your code, you gain confidence in that the file will always be closed correctly, even if you forget to manually close it, or if an exception is raised before you do.

```python
with open(...) as handle:
    ...
```
is essentially the same as
```python
handle = open(...)
try:
    ...
finally:
    handle.close()
```

**Comment by munkm** [comment]
> FWIW I think that actual test coverage for this is required if we're going to merge it before the yt 4.0 release

As discussed in the development slack, the test coverage will come as a follow up PR from @brittonsmith with data included.

**Review by brittonsmith — APPROVED** [review]

**Comment by neutrinoceros** [comment]
Depending on how you guys want to approach merging this and the companion PR, consider you have my approval to move forward. I would still prefer to see some test right away but I don't want to be a blocker

File: yt/frontends/enzo/data_structures.py
Comment (brittonsmith) [inline yt/frontends/enzo/data_structures.py:390]: This needs to get the shape instead of the value.
```suggestion
                            self.io._array_fields[field] = (node[ptype][field].shape[1:],)
```

**Comment by munkm** [comment]
Woooo @fearmayo! Really great job getting this PR over the line. We really appreciate your patience and the time you put into this. Thank you!

**Pull request merged**
